File: src/0.GettingStarted/RandomWord.cpp
/**
 * A program that reads a sequence of words from standard input and prints one of those
 * words uniformly at random. This implementation uses reservoir sampling algorithm to
 * ensure each word has an equal probability of being selected.
 *
 * Compilation: g++ -o build/RandomWord src/0.GettingStarted/RandomWord.cpp
 * Execution: build\RandomWord.exe < data\animals8.txt
 * Get-Content .\data\animals8.txt | .\build\RandomWord.exe 
 */

#include <iostream>
#include <string>
#include <random>

int main() {
    // Champion holds the currently selected word based on reservoir sampling
    std::string champion;
    
    // i represents the count of words processed so far
    int i = 1;
    
    // Create random number generator with a random seed
    std::random_device rd;
    std::mt19937 gen(rd());
    
    std::string word;
    // Process each word from standard input using reservoir sampling
    while (std::cin >> word) {
        // For the i-th word, select it as the new champion with probability 1/i
        std::bernoulli_distribution dist(1.0 / i);
        if (dist(gen)) {
            champion = word;
        }
        i++;
    }
    
    // Output the final selected word
    std::cout << champion << std::endl;
    return 0;
}

File: src/0.GettingStarted/RandomWord.py
"""
This program reads in a sequence of words from standard input and prints one of those words uniformly at random.
It uses reservoir sampling to select the word, which ensures that when the program reads
the nth word, it selects each of the n words seen so far with probability 1/n.
"""

import sys
import random

def main():
    """
    Main function that implements the reservoir sampling algorithm to select a random word.
    Reads words from standard input and prints the selected champion word.
    """
    champion = ""
    i = 1
    
    for line in sys.stdin:
        for word in line.split():
            if random.random() < 1.0 / i:
                champion = word
            i += 1
    
    print(champion)

if __name__ == "__main__":
    main()